docker/Dockerfile

```diff
@@ -0,0 +1,182 @@
+# Copyright 2019 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ------------------------------------------------------------------------------
+
+# Description:
+#   Builds the environment needed to build Avalon shell.
+#
+#  Configuration (build) parameters
+#  - proxy configuration: https_proxy http_proxy ftp_proxy
+#
+# Build:
+#   $ docker build docker -f docker/Dockerfile.tcf-dev -t avalon-shell-dev
+#   if behind a proxy, you might want to add also below options
+#   --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy --build-arg ftp_proxy=$ftp_proxy
+#
+# Run:
+#   $ cd avalon
+#   $ docker run -v $(pwd):/project/avalon -it avalon-shell-dev /bin/bash
+#   (to run with SGX HW, add options '--device=/dev/isgx -v /var/run/aesmd:/var/run/aesmd ')
+#   make clean && make
+
+
+FROM ubuntu:bionic as base_image
+
+# Ignore timezone prompt in apt
+ENV DEBIAN_FRONTEND=noninteractive
+
+# Add necessary packages
+RUN apt-get update \
+ && apt-get install -y -q \
+    build-essential \
+    ca-certificates \
+    pkg-config \
+    python3-toml \
+    python3-requests \
+    python3-colorlog \
+    python3-twisted \
+    software-properties-common \
+ && apt-get -y -q upgrade \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+# Make Python3 default
+RUN ln -sf /usr/bin/python3 /usr/bin/python
+
+# Installing solidity compiler
+RUN add-apt-repository ppa:ethereum/ethereum \
+   && apt-get update \
+   && apt-get install solc
+
+#Build Avalon Enclave Manager intermediate docker image
+FROM base_image as build_image
+
+RUN apt-get update \
+ && apt-get install -y -q \
+    cmake \
+    make \
+    libprotobuf-dev \
+    python3-dev \
+    python3-pip \
+    swig \
+    software-properties-common \
+    wget \
+    tar
+
+# Install setuptools,py-solc-x and web3 packages using pip because
+# these are not available in apt repository.
+# Install nose2 for running tests.
+RUN pip3 install --upgrade setuptools json-rpc web3 py-solc-x nose2
+
+WORKDIR /tmp
+# Build ("Untrusted") OpenSSL, then build ("trusted") SGX OpenSSL
+# Note: The latter will compile in HW or SIM mode depending on the
+# availability of /dev/isgx and /var/run/aesmd/aesm.socket
+RUN OPENSSL_VER=1.1.1d \
```

Comment:
OpenSSL is built in many components. We should try creating a separate docker image for openSSL and reuse it in other components to reduce overall build time for Avalon. Could be done in a separate PR as well

Reply:
This change need to be made in separate PR.